tests/usb_interfaces.test.cpp

```diff
@@ -0,0 +1,737 @@
+    return false;
+  }
+
+public:
+  mock_iface_descriptor m_iface_one;
+  std::string_view m_name_one;
+  mock_iface_descriptor m_iface_two;
+  std::string_view m_name_two;
+  bool m_wrote_descriptors = false;
+};
+// NOLINTEND
+
+}  // namespace
+
+// Test the usb highlevel interface (interface descriptor level)
+boost::ut::suite<"usb_iterface_req_bitmap_test"> req_bitmap_test = []() {
+  using namespace boost::ut;
+  using bitmap = usb_interface::req_bitmap;
+
+  "req bitmap construction from byte test"_test = []() {
+    u8 raw_byte = 0b10000001;
+    bitmap bm(raw_byte);
+
+    // Test recipient
+    expect(that % static_cast<bool>(bitmap::recipient::interface ==
+                                    bm.get_recipient()));
+
+    // Test type
+    expect(that % static_cast<bool>(bitmap::type::standard == bm.get_type()));
+
+    // Test direction
+    expect(that % true == bm.is_device_to_host());
+  };
+
+  "req bitmap construction from enums test"_test = []() {
+    bitmap bm(false, bitmap::type::class_t, bitmap::recipient::endpoint);
+    u8 raw = bm.m_bitmap;
+
+    // Test direction
+    expect(that % false == bool(raw & 1 << 7));
+
+    // Test type
+    expect(that % 1 == ((raw & 0b11 << 5) >> 5));
+
+    // Test recipient
+    expect(that % 2 == (raw & 0xF));
+  };
+};
+
+// Test the usb highlevel interface (interface descriptor level)
+boost::ut::suite<"usb_interface_test"> usb_interface_test = []() {
+  using namespace boost::ut;
+  using descriptor_delta = usb_interface::descriptor_delta;
+  using req_bitmap = usb_interface::req_bitmap;
+
+  mock_usb_endpoint ep;
+  ep.m_info = usb_endpoint_info{ .size = 8, .number = 1, .stalled = false };
+
+  mock_interface iface("Mock", std::move(ep));
+
+  "usb_interface write descriptor test"_test = [iface]() mutable {
+    std::array<byte const, 9> desc = {
+      iface_desc_length,
+      iface_desc_type,
+      0,  // bInterfaceNumber
+      0,  // bAlternateSetting
+      1,  //  bNumEndpoints
+      2,  //  bInterfaceClass
+      3,  // bInterfaceSubClass
+      4,  // bInterfaceProtocol
+      1   // iInterface
+    };
+    auto intended_byte_stream_arr = make_scatter_bytes(desc);
+    scatter_span<byte const> intended_byte_stream(intended_byte_stream_arr);
+    auto f = [intended_byte_stream](scatter_span<hal::byte const> p_data) {
+      // std::ignore = p_data;
+      std::ignore = intended_byte_stream;
+      // Verify size is correct
+      expect(that % iface_desc_length == scatter_span_size(p_data));
+
+      // Verify content is the same
+      expect(that % (intended_byte_stream == p_data));
+    };
+    auto delta = iface.write_descriptors(f, 0, 1);
+
+    expect(descriptor_delta{ 1, 1 } == delta);
+  };
+
+  "usb_interface write string descriptor test"_test = [iface]() mutable {
+    std::ignore = iface.write_descriptors(
+      [](scatter_span<hal::byte const> p_data) { std::ignore = p_data; }, 0, 1);
+    std::string expected_iface_name = "Mock";
```

Comment:
Just a thought: Should we make this string_u16 like it is in USB or just say, "nah, the bytes come over, thats all that matters". I think the later is fine.